assets/src/scripts/_utils.js

```diff
@@ -1,37 +1,43 @@
+import htm from "htm";
+import Papa from "papaparse";
+import vhtml from "vhtml";
+import { csvData } from "./_signals";
+
+export const html = htm.bind(vhtml);
+
 export const getFileExt = (str) => str.split(`.`).pop();
 
 export function csvStringToTable(csvString, el) {
-  const rows = csvString.trim().split(/\r?\n|\r/); // Regex to split/separate the CSV rows
-  let table = ``;
-  let tableRows = ``;
-  let tableHeader = ``;
-
-  rows.forEach((row, rowIndex) => {
-    let tableColumns = ``;
-    const columns = row.split(`,`); // split/separate the columns in a row
-    columns.forEach((column) => {
-      tableColumns +=
-        rowIndex === 0
-          ? `<th class="p-1">${column}</th>`
-          : `<td class="p-1">${column}</td>`;
-    });
-    if (rowIndex === 0) {
-      tableHeader += `<tr>${tableColumns}</tr>`;
-    } else {
-      tableRows += `<tr class="divide-x divide-gray-200 odd:bg-gray-50">${tableColumns}</tr>`;
-    }
-  });
-
-  table += `<table class="min-w-full divide-y divide-gray-300 text-left text-sm text-gray-900">`;
-  table += `<thead class="font-semibold bg-gray-200">`;
-  table += tableHeader;
-  table += `</thead>`;
-  table += `<tbody class="divide-y divide-gray-200">`;
-  table += tableRows;
-  table += `</tbody>`;
-  table += `</table>`;
-
-  el.innerHTML = table; // eslint-disable-line
+  const csvToJson = Papa.parse(csvString).data;
+  csvData.value = csvToJson;
+
+  const bodyCell = (row) =>
+    row.map((cell) => html`<td class="p-1">${cell}</td>`);
+  const bodyRows = csvData.value.map((row, i) =>
+    i > 0
+      ? html`<tr class="divide-x divide-gray-200 odd:bg-gray-50">
+          ${bodyCell(row)}
+        </tr>`
+      : ``
+  );
+
+  const table = html`
+    <table
+      id="csvTable"
+      class="min-w-full divide-y divide-gray-300 text-left text-sm text-gray-900"
+    >
+      <thead class="font-semibold bg-gray-200">
+        <tr>
+          ${csvData.value[0].map((cell) => html`<th>${cell}</th>`)}
```

Comment:
Rather than later indexing a json object (and thus need signal), its it possibly to do that here and we just render table and highlighting in one go?

I ask because this current approach of a global signal and two render passes is not going to scale nicely to multiple csv files for one output?

Reply:
For example, if we parsed the supression data python side, you should have all you need at this point to render the entire table?

Reply:
That is a really good point, I'm going to land a big refactor which should address some of this 🚀